### PR TITLE
docs: Use yaml build config for vars

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Use yaml build config for vars
Additional Description:

Another breakout from #15229 

This adds a build configuration file for vars passed through to sphinx

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: docs/_ext/validating_code_block.py [hunk 1]
```diff
@@ -22,28 +25,58 @@ class ValidatingCodeBlock(CodeBlock):
...
+
+    @property
+    def validator_args(self) -> tuple:
+        args = (
+            self.options.get('type-name'),
+            '-s',
+            '\n'.join(self.content),
+        )
+        return (
+            args + ("--descriptor_path", self.descriptor_path) if self.descriptor_path else args)
```
Comment (dmitri-d) [inline docs/_ext/validating_code_block.py:56]: I'm not sure I understand how this supposed to work -- this potentially comes from a global config file, but is fragment specific? It also appears that the config file is being overwritten in `docs/build.sh`, without a way to define the descriptor. Do we need this parameter at all?

**Comment by dmitri-d** [comment]
I think this looks good, other than the question re: `descriptor_path` parameter in the validating code block.

File: docs/_ext/validating_code_block.py (hunk 1 above)
Comment (phlax) [reply docs/_ext/validating_code_block.py:56]: yep - in #15229 the sphinx job is moved to a bazel `build` job

in this case the path to the descriptor file is different to the one set in `config_validation` (which can now accept this parameter) and so bazel passes the `location` via the config file here

Comment (dmitri-d) [reply docs/_ext/validating_code_block.py:56]: To see if I got it: 
 * the descriptor_path isn't fragment-specific, but if differs from the hard-coded one can be overriden
  * in this case the descriptor path needs to be overriden
  
  I don't see where in the config file the descriptor_path is being defined however, can you point me to it pls?

Comment (phlax) [reply docs/_ext/validating_code_block.py:56]: in this pr the job has not yet been moved to bazel build so it isnt

here it is tho https://github.com/envoyproxy/envoy/pull/15229/files#diff-92ae908436b530a281eddfb50321c5a055ad4fa90e0194f0651b58ac39341800R206 and this is then written to the config file here https://github.com/envoyproxy/envoy/pull/15229/files#diff-b7bc9b4ee0614d05a36dbf7391fe7c20bc20f0f0a3dc67b86eb859139a333aa9R50  and consumed from there

File: docs/_ext/validating_code_block.py [hunk 2]
```diff
@@ -22,28 +25,58 @@ class ValidatingCodeBlock(CodeBlock):
...
     option_spec.update(CodeBlock.option_spec)
-    skip_validation = (os.getenv('SPHINX_SKIP_CONFIG_VALIDATION') or 'false').lower() == 'true'
+
+    @cached_property
+    def configs(self) -> dict:
+        _configs = dict(
+            descriptor_path="",
+            skip_validation=False,
+            validator_path="bazel-bin/tools/config_validation/validate_fragment")
+        if os.environ.get("ENVOY_DOCS_BUILD_CONFIG"):
```
Comment (dmitri-d) [inline docs/_ext/validating_code_block.py:35]: Is writing to a file on access a common pattern for a `@[cached_]property`? Would it be possible to replace this with an explicit call?

Comment (phlax) [reply docs/_ext/validating_code_block.py:35]: its not *writing* a file its *reading* it and caching the contents as a property of the class

there isnt so much a common pattern with `cached_property` as it has relatively recently been added to python (from django)

certainly you wouldnt want to do this with a not cached property, and probably not for something that actually persists

Comment (phlax) [reply docs/_ext/validating_code_block.py:35]: > there isnt so much a common pattern with...

altho in so far as there is one - this is it - ie preventing repeated `io` calls for a class property

> probably you wouldnt want to do this with a not cached property,

in this case you would want generally want to be explicit - ie every time i call this, get me the current value

> and probably not for something that actually persists

i think this is more moot - eg get me a file path - but the first time you do - please populate it with `x` contents

Comment (dmitri-d) [reply docs/_ext/validating_code_block.py:35]: Ugh, indeed, I was thinking about https://github.com/envoyproxy/envoy/pull/15229/files#diff-b7bc9b4ee0614d05a36dbf7391fe7c20bc20f0f0a3dc67b86eb859139a333aa9R50 when writing the comment. Makes sense here, never mind me.

**Comment by dmitri-d** [comment]
lgtm

**Review by dmitri-d — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Pull request merged**